src/decompiler-ast.inl

```diff
@@ -204,41 +206,102 @@ struct AST {
-    auto start = stack.size();
-    auto stack_depth_start = stack_depth;
+    auto start = exp_stack.size();
+    auto wasm_stack_depth_start = wasm_stack_depth;
+    auto wasm_stack_in_variables = wasm_stack_depth;
     bool unreachable = false;
     for (auto& e : es) {
       Construct(e);
       auto arity = mc.GetExprArity(e);
-      stack_depth -= arity.nargs;
-      stack_depth += arity.nreturns;
+      wasm_stack_depth -= arity.nargs;
+      wasm_stack_in_variables = std::min(wasm_stack_in_variables,
+                                         wasm_stack_depth);
       unreachable = unreachable || arity.unreachable;
-      assert(unreachable || stack_depth >= 0);
-      if (arity.nreturns > 1) {
-        // Multivalue: we "push" everything on to the stack.
-        NewNode(NodeType::PushAll, ExprType::Nop, nullptr, 1);
-        // All values become pops.
-        for (Index i = 0; i < arity.nreturns; i++)
-          NewNode(NodeType::Pop, ExprType::Nop, nullptr, 0);
-        // TODO: can also implement a push_all_but_one that returns the top,
-        // then insert N-1 pops below it. Or have a function that returns N
-        // values direcly to N arguments for when structs are passed on,
-        // etc.
+      assert(unreachable || wasm_stack_depth >= wasm_stack_depth_start);
+      wasm_stack_depth += arity.nreturns;
+      // We maintain the invariant that a wasm_stack_depth of N is represented
+      // by the last N exp_stack items (each of which returning exactly 1
+      // value), all exp_stack items before it return void ("statements").
+      // In particular for the wasm stack:
+      // - The values between 0 and wasm_stack_depth_start are part of the
+      //   parent block, and not touched here.
+      // - The values from there up to wasm_stack_in_variables are variables
+      //   to be used, representing previous statements that flushed the
+      //   stack into variables.
+      // - Values on top of that up to wasm_stack_depth are exps returning
+      //   a single value.
+      // The code below maintains the above invariants. With this in place
+      // code "falls into place" the way you expect it.
+      auto num_vars = wasm_stack_in_variables - wasm_stack_depth_start;
+      auto num_vals = wasm_stack_depth - wasm_stack_in_variables;
+      if (arity.nreturns == 0 && wasm_stack_depth > wasm_stack_depth_start) {
+        // We have a void item on top of the exp_stack, so we must "lift" the
+        // previous values around it.
+        // We track what part of the stack is in variables to avoid doing
+        // this unnecessarily.
+        if (num_vals > 0) {
+          // We have actual new values that need lifting.
+          // This puts the part of the stack that wasn't already a variable
+          // (before the current void exp) into a FlushToVars.
+          // Now re-generate these values after the void exp as vars.
+          FlushVars(num_vals, num_vars, 1, num_vals);
```

Comment:
As far as I can tell this is the only use of non-0 dip? That feels like a disproportionate amount of complexity/machinery for something with one use. Is dip the right abstraction going forward, or should we special-case this flush?

Reply:
I could remove the `dip` arg by making only that case pop the stack..